onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu

```diff
@@ -1315,6 +1315,81 @@
                                             int chunk_size,
                                             cudaStream_t stream);
 
+constexpr unsigned int kTileSize = 32;
+constexpr unsigned int kSeqTileSize = 16;
+
+template <typename T>
+__global__ void ReorderPastStatesKernel(T* out_buffer,
+                                        const T* in_buffer,
+                                        int batch_size,
+                                        int num_heads,
+                                        int max_length,
+                                        int chunked_head_size,
+                                        int equv_chunk_size) {
+  //[B, N, max_length, H2(head_size/chunk_size), chunk_size] -> [B, N, H2(head_size/chunk_size), max_length, chunk_size]
+  __shared__ T tile[kSeqTileSize][kTileSize + 1];
+
+  const int b = blockIdx.x;
+  const int n = blockIdx.y;
+  const int s = blockIdx.z * kSeqTileSize + threadIdx.x;
+  const int h2 = threadIdx.y;
+  const int c = threadIdx.z;
+
+  if (s < max_length) {
+    const int in_offset = b * num_heads * max_length * chunked_head_size * equv_chunk_size +
+                          n * max_length * chunked_head_size * equv_chunk_size +
+                          s * chunked_head_size * equv_chunk_size +
+                          h2 * equv_chunk_size +
+                          c;
+
+    const int tile_x = threadIdx.x;
+    const int tile_y = threadIdx.y * equv_chunk_size + threadIdx.z;
+    tile[tile_x][tile_y] = in_buffer[in_offset];
+
+    __syncthreads();
+
+    const int out_offset = b * num_heads * max_length * chunked_head_size * equv_chunk_size +
+                           n * max_length * chunked_head_size * equv_chunk_size +
+                           h2 * max_length * equv_chunk_size +
+                           s * equv_chunk_size +
+                           c;
+
+    out_buffer[out_offset] = tile[tile_x][tile_y];
+  }
+}
+
+void ReorderPastStatesKernelLauncher(void* out_buffer,
+                                     const void* in_buffer,
+                                     int batch_size,
+                                     int num_heads,
+                                     int max_length,
+                                     int head_size,
+                                     int chunk_size,
+                                     cudaStream_t stream) {
+  const int chunked_head_size = head_size / chunk_size;
+  const dim3 block(16, chunked_head_size, chunk_size / 4);
+  const dim3 grid(batch_size, num_heads, (max_length + block.x - 1) / block.x);
+  if (chunk_size == 4) {
```

Comment:
> if (chunk_size == 4) { [](http://example.com/codeflow?start=1&length=23)

there is no need to differentiate half or float. The last dimension keeps 16bytes, which is a float4. So you can convert a half tensor of [B, N, S, H] to float4 tensor with shape [B, N, S, H/8] and a float tensor to float4 tensor with shape [B, N, S, H/4]